### PR TITLE
Arch 02 cmake proposal

### PR DESCRIPTION
**Pull request by mBialczak**
1. Structure for current backend structure with cmake files proposal.
2. Adjustment of current backend files for compilation and linking purposes.
3. Basic test structure with cmake files proposed. Not perfect yet.
4. In main cmake file rule for downloading and building fixed protobuf lib proposed.
5. Basic construction test of all existing backend structures done, to test if it works.

General solution testing instructions (I will have to come up with more detailed ones sorry for shorthand)
1. Building backend with installation of built backend libs and includes to seperate folders. (this will create two folders: external_libraries - currently for protobuf and own_libraries - for our libs built from backend).
CAUTION: downloading and building protobuf specific version can take quite a long time - but once only.
```
cd build
cmake -DCMAKE_BUILD_TYPE=Debug ../modules/backend/
cmake --build . --parallel --target install
```
2. Building test target (after building backend)
```
cd build_tests
cmake -DCMAKE_BUILD_TYPE=Debug ../modules/tests/
cmake --build . --parallel
```
3. New test can be run in normal manner. It goes to bin folder.

### CODE REVIEW

File: modules/backend/clinic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+set(CLINIC_SOURCES
+    ${CMAKE_CURRENT_SOURCE_DIR}/Clinic.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/room/Room.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/staff/doctor/Doctor.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/staff/Staff.cpp
+)
+
+add_library(clinic-lib STATIC ${CLINIC_SOURCES})
+
+target_compile_options(clinic-lib PRIVATE -Werror -Wall -Wextra -pedantic)
```
Comment (fbtom) [inline modules/backend/clinic/CMakeLists.txt:10]: What do you think about becoming more strict in production code?
For example those could be added:
` -Wunnused -Wshadow -Wold-style-cast -Wconversion -Wdangling-pointer`

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: modules/backend/clinic/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+target_include_directories(clinic-lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
+
+# install created lib in our own libraries directory
+install(TARGETS clinic-lib
+        ARCHIVE DESTINATION ${OWN_LIBRARIES_DIR})
+
+# copy include files to our own libraries directory
+install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/ DESTINATION ${OWN_LIBRARIES_DIR}/include
+        FILES_MATCHING PATTERN "*.h" PATTERN "*.hpp")
```
Comment (fbtom) [inline modules/backend/clinic/CMakeLists.txt:25]: Please clarify me if I am wrong.

1st install: installs clinic library as archive to own_lib_dir
2nd install: copy header files to own_lib_dir/include 

Why is that so? I am trying to understand purpose, pros and cons

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: modules/backend/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,37 @@
+cmake_minimum_required(VERSION 3.23)
+project(desktop-business-app-backend)
+
+set(CMAKE_CXX_STANDARD 20)
+
+#TODO: adjust both when we have final version of structure of our project
+set(EXTERNAL_LIBRARIES_DIR ${PROJECT_SOURCE_DIR}/../../external_libraries/)
```
Comment (fbtom) [inline modules/backend/CMakeLists.txt:7]: Is  PROJECT_SOURCE_DIR a main level -> '/ '?
If so, then I understand why /../../ are necessary :)

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: modules/backend/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+set(RELEASE_TAG "v21.0")
+set(PROTOBUF_VERSION "3.21.0")
+set(PROTOBUF_PREFIX "${EXTERNAL_LIBRARIES_DIR}/protobuf")
+ExternalProject_Add(protobuf
+    URL "${PROTOBUF_BASE_URL}/${RELEASE_TAG}/protobuf-cpp-${PROTOBUF_VERSION}.tar.gz"
+    PREFIX ${PROTOBUF_PREFIX}
+    CONFIGURE_COMMAND <SOURCE_DIR>/configure --prefix=<INSTALL_DIR>
+    BUILD_COMMAND ${MAKE}
+    DOWNLOAD_EXTRACT_TIMESTAMP true
+)
```
Comment (fbtom) [inline modules/backend/CMakeLists.txt:25]: [Q] Is it possible to hide it in a different cmake and call it just like a function?

File: modules/backend/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+set(CMAKE_PREFIX_PATH ${CMAKE_PREFIX_PATH} ${PROTOBUF_PREFIX})
+
+find_package(Protobuf REQUIRED)
```
Comment (fbtom) [inline modules/backend/CMakeLists.txt:29]: As we specify protobuf_version above, wouldn't be that good to also use it here as an arugment, to make sure it's correct?

File: modules/backend/clinic/Clinic.cpp [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#include "Clinic.hpp"
+
+#include "room/iRoom.hpp"
+#include "staff/Staff.hpp"
```
Comment (fbtom) [inline modules/backend/clinic/Clinic.cpp:4]: Do I understand it correctly, that because of CMake with CMAKE_CURRENT_SOURCE_DIR under `modules/backend/clinic` it is not necessary to use  `modules/backend/clinic` in the includes? 
Or it could be used? Or it will look weird like  `../../../modules/backend/clinic/'?

File: modules/backend/clinic/Clinic.hpp [hunk 7]
```diff
@@ -1,23 +1,28 @@
...
 
+namespace staff
+{
+class Staff;
+}
+
+namespace room
+{
+class iRoom;
+}
```
Comment (fbtom) [inline modules/backend/clinic/Clinic.hpp:17]: Why are these necessary? 
Interfaces headers are not heavy, they can be included here, in the header. This is the first time I am seeing such thing tbh

File: modules/backend/common/enumerations/sex/Sex.cpp [hunk 8]
```diff
@@ -21,7 +11,7 @@ std::string toString(const Sex& sex)
     case Sex::Female:
         return "Female";
     default:
-        return;
+        return {"Unknown"};
```
Comment (fbtom) [inline modules/backend/common/enumerations/sex/Sex.cpp:14]: Such situation will never occur if Pesel will be responsible for defining Sex, we can simply keep 'return;'

Comment (fbtom) [reply modules/backend/common/enumerations/sex/Sex.cpp:14]: Or if it's necessary add Unknown as a state of enum type and some convertion layer that will change values different than Female/Male to Unknown

File: modules/backend/patients/patient/payments/Payments.hpp [hunk 9]
```diff
@@ -7,6 +7,7 @@ namespace patient
 
 struct Payments
 {
+    // TODO: consider type: float or double maybe?
     std::uint8_t paid_;
     std::uint8_t planned_;
     std::uint8_t overdue_;
```
Comment (fbtom) [inline modules/backend/patients/patient/payments/Payments.hpp:13]: From my pov we could use integer types and simply move two digits. 
This way we will omit possible issues with floating types.

File: modules/backend/patients/patient/treatments/Treatments.cpp [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#include "Treatments.hpp"
+
+namespace patient
+{
+
+void Treatments::updateTreatment([[maybe_unused]] const common::TreatmentState treatment_state)
+{
+    // move Treatment from planned to done
```
Comment (fbtom) [inline modules/backend/patients/patient/treatments/Treatments.cpp:8]: We should mark it with [#128](https://github.com/coders-school/desktop-business-app/issues/128). As there should be more states and logic. 
Something like "state machine" should be created

File: modules/backend/visits/VisitCollection.hpp [hunk 11]
```diff
@@ -1,25 +1,24 @@
...
-    {
-        visits_.emplace_back(visit);
-    }
+    void appendVisit(const std::shared_ptr<iVisit>& visit);
+    const std::vector<std::shared_ptr<iVisit>>& getVisits() const;
+
+  private:
+    //  TODO: type of pointer to be decided here (unique/shared/raw).
+    //  Functions require this consideration too then
+    std::vector<std::shared_ptr<iVisit>> visits_;
```
Comment (fbtom) [inline modules/backend/visits/VisitCollection.hpp:20]: Unique ptrs.
Raw pointers should be disallowed.
We need to add a rule, that raw pointers can be only used, when interface of external library has it, then it can be used, but still we should make it a safe pointer after obtaining such.

File: modules/tests/integration-tests/basic_construction_tests/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+                                              gtest_main
+                                              ${CLINIC_LIB}
+                                              ${COMMON_LIB}
+                                              ${PATIENTS_LIB}
+                                              ${SERIALIZER_LIB}
+                                              ${VISITS_LIB}
+                                              ${WAREHOUSE_LIB}
+)
+
+gtest_discover_tests(basic_construction-test)
```
Comment (fbtom) [inline modules/tests/integration-tests/basic_construction_tests/CMakeLists.txt:20]: Looks clean, good job

File: modules/backend/clinic/CMakeLists.txt (hunk 1 above)
Comment (mBialczak) [reply modules/backend/clinic/CMakeLists.txt:10]: Good idea, I will add it.

File: modules/tests/integration-tests/basic_construction_tests/basic_construction_test.cpp [hunk 13]
```diff
@@ -0,0 +1,200 @@
+#include <gtest/gtest.h>
```
Comment (fbtom) [inline modules/tests/integration-tests/basic_construction_tests/basic_construction_test.cpp:1]: directory name integration-tests needs to be changed to integration_tests

Except for that, those tests needs to be separated respectively. it is not necessary to do it now

File: modules/backend/clinic/CMakeLists.txt (hunk 2 above)
Comment (mBialczak) [reply modules/backend/clinic/CMakeLists.txt:25]: Yes thats correct.
First install install binary libraries. The second - copies headers.
Headers are necessary, as in targets using libs any include without them would not work. I was surprised too.
This was suggested by copilot.
But i tested it anyway as first the headers were not installed - and none include in tests worked.
In the first approach I had one backend lib. But as I saw the problem with includes, I decided for seperate libs.

All in all if we want one backend lib - I think that it will be possible but install of includes most likely will have to stay.
Unless we find a better solution of course.

It's hard to find pros and cons so far as I see no other solution at this point. Perhaps there are some package managers, which could make something better.
I wonder as well, if we add header files to sources together with .cpp maybe they would also be included in lib it self.

**Review by BoskiJanusz — APPROVED** [review]
chapeau bas

File: modules/backend/CMakeLists.txt (hunk 3 above)
Comment (mBialczak) [reply modules/backend/CMakeLists.txt:7]: PROJECT_SOURCE_DIR if I understood copilot right, is the main cmake level dir (or perhaps where PROJECT_NAME var is established. We will have to modify it when we get into final project structure.
In this example it is backend module dir if I got it right

File: modules/backend/CMakeLists.txt (hunk 4 above)
Comment (mBialczak) [reply modules/backend/CMakeLists.txt:25]: Perfect question and good idea.
Yes, it can be extracted and solution is simple. I will add it shortly.

File: modules/backend/clinic/Clinic.cpp (hunk 6 above)
Comment (mBialczak) [reply modules/backend/clinic/Clinic.cpp:4]: Yes. it is like that.
In general, In current solution I made every module top level dir accesible. And including anything inside module directly resembles it's own structure.
For ex. if we are in Clinic, we simply include room/iRoom.hpp because this is folder structure.

Conversely, if we are in different module like werehouse and in Warehouse.hpp (or even deeper), if we want something from clinic, like room, we are only interested in top level dir of clinic (which is implicitly provided).
So there we only write #include "room/iRoom.hpp" as well.

module/backend/clinic in general no longer works.
Absolute paths should work (but no sense). Relative paths will most likely work as well

Comment (mBialczak) [reply modules/backend/clinic/Clinic.cpp:4]: So we only need to remember where top module dirs are.
This could be easily seen in installed headers in own_libraries
image

File: modules/backend/clinic/Clinic.hpp (hunk 7 above)
Comment (mBialczak) [reply modules/backend/clinic/Clinic.hpp:17]: Forward declarations here are necessary after I moved includes to .cpp files.

I try to stick to the recommendation to include only what is necessary and optimally - in cpp files only.
It is hard to say which is better, as conflicting opinions exist.
In some cases it could speedup compilation. Hard to observe now but maybe as the project grows...

If we decide to resign from forward declarations in favor of includes no great harm will come in our case I think.
But it should be deliberate decision, which we might include in our guidelines.
Some little research for pros and cons might be good beforehand.

File: modules/backend/patients/patient/payments/Payments.hpp (hunk 9 above)
Comment (mBialczak) [reply modules/backend/patients/patient/payments/Payments.hpp:13]: maybe. I don't understand exactly. Oh I see: for example 100 for 1 zł for example. Good idea.
But any way (and especially if so) we must provide bigger type. If we stick with uint8_t we are constrained to 256.
If we decide to apply 100 for 1 rule this will mean maximium of 2,56 (currency) in our code.

File: modules/backend/patients/patient/treatments/Treatments.cpp (hunk 10 above)
Comment (mBialczak) [reply modules/backend/patients/patient/treatments/Treatments.cpp:8]: I will mark it as recommended.
For state machine I don't want to respond yet without #128 problem considerations.
It might be good for this case, but hard to tell without further stydying the problem.

File: modules/backend/visits/VisitCollection.hpp (hunk 11 above)
Comment (mBialczak) [reply modules/backend/visits/VisitCollection.hpp:20]: Ok, it's good to decide early.
I will change to unique_ptr.

But I don't aggree to the rule exactly.
Because here by using unique_ptr we force on raw pointers and reference in our code as well.
From visit collection we will only return raw pointers and references then.
But this is good in my opinion: 
1) single point of managing resource
2) everywhere else in our code we can only use raw ptr or ref without bothering about object lifecycle.

Comment (mBialczak) [reply modules/backend/visits/VisitCollection.hpp:20]: So we can add recommendation like this:
"where ever possible to establish single point of resource managment,use unique_ptr and return raw pointers or references from getters".
Or something like that

File: modules/tests/integration-tests/basic_construction_tests/basic_construction_test.cpp (hunk 13 above)
Comment (mBialczak) [reply modules/tests/integration-tests/basic_construction_tests/basic_construction_test.cpp:1]: Yeap. changed the name locally to integration_tests and unit_tests. Will land in next commit pack.

In general this basic_test is doomed to be removed as soon as unit_tests will exist.
So I think it is best to remove specific part of test as soon as unit_test is started, until file is empty and can be removed.

File: modules/backend/patients/patient/payments/Payments.hpp (hunk 9 above)
Comment (fbtom) [reply modules/backend/patients/patient/payments/Payments.hpp:13]: Yes, we should go for uint32_t probably. It should be more than enough.
But that's for the future

File: modules/backend/clinic/CMakeLists.txt (hunk 1 above)
Comment (mBialczak) [reply modules/backend/clinic/CMakeLists.txt]: Done

File: modules/backend/CMakeLists.txt (hunk 4 above)
Comment (mBialczak) [reply modules/backend/CMakeLists.txt]: Done extraction

File: modules/backend/common/enumerations/sex/Sex.cpp (hunk 8 above)
Comment (mBialczak) [reply modules/backend/common/enumerations/sex/Sex.cpp]: Done

File: modules/tests/integration-tests/basic_construction_tests/basic_construction_test.cpp (hunk 13 above)
Comment (mBialczak) [reply modules/tests/integration-tests/basic_construction_tests/basic_construction_test.cpp]: DONE

File: modules/backend/CMakeLists.txt (hunk 5 above)
Comment (mBialczak) [reply modules/backend/CMakeLists.txt]: I have set required version to 3.21.0
I have some doubts if it will work for clean external_lib installation but if anybody reports problem it will be easily fixed.
DONE

**Review by fbtom — APPROVED** [review]

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @mBialczak for making course materials better!

**Comment by ziobron** [comment]
Thanks @BoskiJanusz for Code Review.
🏅 1 XP granted
